lib/server.js

```diff
@@ -61,8 +61,8 @@ function createServer (options, httpHandler) {
     if (Object.prototype.hasOwnProperty.call(listenOptions, 'host') === false) {
       listenOptions.host = host
     }
-
-    if (host === 'localhost') {
+    const serverFactoryListening = (options.serverFactory && options.serverFactory.listening) || (options.serverFactory === server)
```

Comment:
`options.serverFactory` is a `function`
Your check will never fulfill.

Reply:
I realized `server` was from `serverFactory` itself so calling it again makes no sense. How about this?
```javascript
const serverFactoryListening = server.listening && options.serverFactory
```

Reply:
It only check the first server, but not the second one.
The idea is https://github.com/fastify/fastify/issues/4730#issuecomment-1535787094

Reply:
1. The first server would throw `error`. - Do you mean `throw new Error()` if the main server is listening?
2. The second server would emit `warning`. - Do you mean `warnings.emit()`?